client/desktop/lib/cookie-auth/index.js

```diff
@@ -70,49 +67,56 @@ function parseCookie( cookieStr ) {
 	}, cookie );
 }
 
-function setSessionCookies( window, onComplete ) {
-	return function ( response ) {
-		let cookieHeaders = response.headers[ 'set-cookie' ];
+function setSessionCookies( window, authorizeResponse ) {
+	return new Promise( ( resolve ) => {
+		let cookieHeaders = authorizeResponse.headers[ 'set-cookie' ];
 		let count = 0;
 		if ( ! Array.isArray( cookieHeaders ) ) {
 			cookieHeaders = [ cookieHeaders ];
 		}
 
 		count = cookieHeaders.length;
 
+		if ( count === 0 ) {
+			return resolve( true );
+		}
+
 		cookieHeaders.map( parseCookie ).forEach( function ( cookie ) {
 			cookie.url = 'https://wordpress.com/';
-			if ( cookie.httponly ) {
+			if ( cookie.HttpOnly || cookie.httpOnly || cookie.httponly ) {
 				cookie.session = true;
 			}
-			window.webContents.session.cookies.set( cookie, function () {
-				count--;
-				if ( count === 0 ) {
-					if ( onComplete ) onComplete();
-				}
-			} );
+			window.webContents.session.cookies.set( cookie );
+			count--;
+			if ( count === 0 ) {
+				return resolve( true );
+			}
```

Comment:
This is strange: the `cookies.set` function is asynchronous and accepts a completion callback. Now this change removes the callback and doesn't wait for `cookies.set` to finish.

Previously, `onComplete` was called only after all the `cookie.set` calls finished. Now everything is called synchronously and it's not even clear why there is a `new Promise()`. There's no async code left.

Reply:
When debugging these boot/auth flows in the Desktop app, I found the nested callbacks really hard to navigate and reason about. I took the liberty of flattening the callbacks used in this file to make things a little easier to grok going forward.

Especially for Main process logic, this is also aligned with a number of Electron APIs which have migrated to using promises instead of callbacks in recent release. Funnily enough, the `cookies.set` call is promise-based as well ([docs](https://www.electronjs.org/docs/api/cookies#cookiessetdetails)) (although seems like there's a callback variant to support legacy usage). Interestingly prefixing with `await` ends up breaking things (awaits indefinitely for some reason).

For the time being will fall back to the callback variant of this method, although makes me a little nervous that things may silently break when the callback API deprecation is final.

Reply:
Spawning several parallel async "threads" in `forEach`, waiting for all of them to finish and resolving then... that's exactly what `Promise.all` does. The `cookieHeaders.forEach` cycle can be rewritten as:
```js
await Promise.all( cookieHeaders.map( async ( cookieStr ) => {
  // ...
} ) );
```

And `setSessionCookies` can become an `async` function.